cpp/src/arrow/util/parallel.h

```diff
@@ -44,6 +45,25 @@ Status ParallelFor(int num_tasks, FUNCTION&& func,
   return st;
 }
 
+template <class FUNCTION, typename T,
+          typename R = typename internal::call_traits::return_type<FUNCTION>::ValueType>
+Future<std::vector<R>> ParallelForAsync(
+    std::vector<T> inputs, FUNCTION&& func,
```

Comment:
Is there a particular reason for taking a vector of input instead of a `num_tasks` such as above?

Reply:
Yes: the callback of the sync version can safely and easily capture references to local stack variables. But the async version cannot. Rather than mess around with wrapping state in a shared_ptr, it's easier to just pass a vector. I can rename this to `ParallelForEachAsync` though to make it clearer.